### ISSUE TITLE
leak-tracker: Support pure Dart package

### ISSUE THREAD

**Issue by Solido** [dart-lang/leak_tracker#257]
Hi,

leak-tracker is an important tool in the dart world as the language is moving from being mainly front to also model rich world logic.

I'm working on a large package that would greatly benefits from memory insights while dealing with deep data structures and operations.

Thanks!

**Comment by polina-c** [comment]
Thanks for filing it! How about design discussion? I sent you message in Twitter to schedule.

**Comment by polina-c** [comment]
Relevant issue: https://github.com/flutter/flutter/issues/161132

**Comment by polina-c** [comment]
Question: can you add dev dependency on flutter_test and use 'testWidgets' instead of 'test' to get leak tracking for your tests?
If this works out, we just need to document this.

**Comment by Solido** [comment]
It should be possible but it require to reconfigure the project from pure dart to flutter as we also need to import the flutter sdk.

**Comment by polina-c** [comment]
Not the entire project, but tests only. And anothe roption is to have separate flutter projext with just tests.
But, yes, ideally we want to update the pure dart method 'test' to support leak tracking.
First step would be to create one page design doc ([template](http://flutter.dev/go/template), [example](http://flutter.dev/go/leak-tracker-make-bots-blocking)) and review it with team.

**Comment by Solido** [comment]
Currently I'm using a monorepo with all logics and tests relying on dart then another flutter project build on the logic. To use flutter test on logic would mean to import Flutter sdk which is not an option for later build purpose.

Flutter projects would not cover the same testing surface thus a case to check for memory leaks first on a pure dart project and then later on widgets.

**Comment by polina-c** [comment]
> Currently I'm using a monorepo with all logics and tests relying on dart then another flutter project build on the logic. To use flutter test on logic would mean to import Flutter sdk which is not an option for later build purpose.
> 
> Flutter projects would not cover the same testing surface thus a case to check for memory leaks first on a pure dart project and then later on widgets.

Makes sense. Then next step is a one-page design doc.

**Comment by polina-c** [comment]
To get attention of Dart test people, it would make sense to file a new issue towards dart test, requesting to integrated `test` with leak tracker the same way as `testWidgets` is integrated: https://github.com/dart-lang/test/issues

**Comment by xvrh** [comment]
@polina-c do you think there is a way to setup leak tracking in Dart tests without direct support from the test package?
Maybe there is something we could put in our `setUp`/`tearDown` methods?

I tried that but without success
```dart
import 'package:leak_tracker/leak_tracker.dart';

void main() {
  setUp(() {
    LeakTracking.start();
  });

  tearDown(() async {
    LeakTracking.declareNotDisposedObjectsAsLeaks();
    final leaks = await LeakTracking.collectLeaks();
    LeakTracking.stop();
    print("${leaks.all}");
  });
```

Do you see something missing?

**Comment by xvrh** [comment]
If I write this:

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:leak_tracker_flutter_testing/leak_tracker_flutter_testing.dart';

void main() {
  LeakTesting.enable();

  testWidgets('Check leak', (tester) async {
    var notifier = ValueNotifier(3);
  });
}
```

And run with `flutter test test/my_flutter_test.dart` I got the error as expected: `Expected: leak free .. Which: contains leaks: ...  ValueNotifier<int>`

If I write this:

```dart
import 'package:flutter/foundation.dart';
import 'package:test/test.dart';
import 'package:leak_tracker/leak_tracker.dart';

void main() {
  setUp(() {
    LeakTracking.start();
  });

  tearDown(() async {
    LeakTracking.declareNotDisposedObjectsAsLeaks();
    final leaks = await LeakTracking.collectLeaks();
    LeakTracking.stop();
    print("${leaks.all}");
  });

  test('Check leak', () async {
    var notifier = ValueNotifier(3);
  });
}
```

And run with the same command `flutter test test/my_dart_test.dart`. `leaks` object is empty.

**Comment by polina-c** [comment]
Thanks for the details.

ValueNotifier is a Flutter object that reports activity (creation and disposal) to FlutterMemoryAllocations, not to leak tracker directly (to see this, check code of ValueNotifier and ChangeNotifier). To connect this activity to leak tracker, you need to run this before starting leak tracker:

```
FlutterMemoryAllocations.instance.addListener(
    (ObjectEvent event) => LeakTracking.dispatchObjectEvent(event.toMap()),
);
```

And you want to replace setup with setupAll, and tearDown with tearDownAll, because leak collection is a heavy operation, that you do not want to happen after every test.

Hope it helps.
